### ISSUE TITLE
Bug: onnx2ncnn doesn't convert "concat" , "constant", "softmax" correctly.

### ISSUE THREAD

**Issue by keyongyu** [Tencent/ncnn#720]
1.  Constant OP:
      given the node like:
        op_type: "Constant"
        attribute {
           name: "value"
           t {
               dims: 2
               data_type: INT64
               raw_data: "\377\377\377\377\377\377\377\377\025\000\000\000\000\000\000\000"
             }
            type: TENSOR
          }
    So, the int64_data will be null, but raw_data exists, onnx2ncnn should process it.
2.  Conversion of "Concat" op is not right.
        axis in "Concat" op is zero based, so, should not be 
                fprintf(pp, " 0=%d", axis-1);

**Comment by nihui** [comment]
for the first issue, the current converter will check raw_data firstly
https://github.com/Tencent/ncnn/blob/b5faa0e519c12a216cfad167058342033fdc5bc9/tools
/onnx/onnx2ncnn.cpp#L174

for the remaining axis issue, onnx use NCHW, while ncnn only support CHW
so the axis 1 actually corresponding to 0 in ncnn

**Comment by kealennieh** [comment]
it seems that the problem still exists in Constant Op.   

TensorProto.raw_data() can get a string, but float_data_size() is 0. Therefore, the value of Constant Op will be empty, which would cause " Unsupported Upsample scales !"  

Any suggestions?

**Comment by keyongyu** [comment]
Yes, the bug is not fixed, please modify the function "get_node_attr_tensor" in onnx2ncnn.cpp


static onnx::TensorProto get_node_attr_tensor(const onnx::NodeProto& node, const char* key)
{
    for (int i=0; i<node.attribute_size(); i++)
    {
        const onnx::AttributeProto& attr = node.attribute(i);
        if (attr.name() == key)
        {
            //return attr.t();
            onnx::TensorProto t= attr.t();
            if(t.has_raw_data()){
                const std::string& raw_data=t.raw_data();
                int element_size=8;
                if(   t.data_type() == onnx::TensorProto_DataType::TensorProto_DataType_UINT64
                   || t.data_type() == onnx::TensorProto_DataType::TensorProto_DataType_INT64
                   )
                    element_size=8;
                else if (t.data_type() == onnx::TensorProto_DataType::TensorProto_DataType_FLOAT
                      || t.data_type() == onnx::TensorProto_DataType::TensorProto_DataType_INT32
                      || t.data_type() == onnx::TensorProto_DataType::TensorProto_DataType_UINT32
                ) {
                    element_size = 4;
                }
                int dim = raw_data.length()/element_size;
                for (int j=0; j<dim; j++) {
                    if(t.data_type() == 7) //int64
                    {
                        const int64_t*  ptr = (const int64_t *) raw_data.c_str();
                        auto elm = ptr[j];
                        t.add_int64_data(elm);
                    }
                    else if (t.data_type() ==1) {
                        const float*  ptr = (const float*) raw_data.c_str();
                        auto elm = ptr[j];
                        t.add_float_data(elm);
                    }
                }
            }
            return t;
        }
    }

    return onnx::TensorProto();
}

**Comment by keyongyu** [comment]
> for the remaining axis issue, onnx use NCHW, while ncnn only support CHW
> so the axis 1 actually corresponding to 0 in ncnn
This is very critical, I exported my model to CHW mode for inference purpose before, now I just export it to 1CHW mode,  all those axis issues are gone now.

**Comment by nihui** [comment]
针对onnx模型转换的各种问题，推荐使用最新的pnnx工具转换到ncnn
In view of various problems in onnx model conversion, it is recommended to use the latest pnnx tool to convert your model to ncnn

```shell
pip install pnnx
pnnx model.onnx inputshape=[1,3,224,224]
```

详细参考文档
Detailed reference documentation
https://github.com/pnnx/pnnx
https://github.com/Tencent/ncnn/wiki/use-ncnn-with-pytorch-or-onnx#how-to-use-pnnx
